### DISCUSSION TITLE
Setting Up TLS for OpenLDAP

### DISCUSSION THREAD

**Discussion by ddspell** [smallstep/certificates#1037] [Q&A]
Using bitnami/openldap.

Any guidance for establishing TLS for OpenLDAP in a container using this application is appreciated.

Thanks!

**Comment by maraino** [comment]
I haven't tried this, but according to https://www.openldap.org/doc/admin24/tls.html you can configure the following directives:
* `TLSCACertificateFile <filename>`: with the CA, the root certificate.
* `TLSCertificateFile <filename>`: with the server certificate.
* `TLSCertificateKeyFile <filename>`: with the server certificate key.

And optionally, `TLSCipherSuite` with a semicolon-separated list of cipher suites. For TLS1.2, good options are `ECDHE-ECDSA-CHACHA20-POLY1305;ECDHE-ECDSA-AES128-GCM-SHA256`. Although I recommend configuring just `TLSProtocolMin` to support only TLSv1.3, I don't know if this is possible, `TLSProtocolMin: 3.3` seems to be TLSv1.2+.

Then depending on how you deploy your docker image, you can use different things. If you use Kubernetes you can either use [autocert](https://github.com/smallstep/autocert/) or cert-manager with [step-issuer](https://github.com/smallstep/step-issuer). But if you are deploying this on docker directly, I would recommend creating a custom image extending `bitnami/openldap` with `step` from [`smallstep/step-cli`](https://hub.docker.com/r/smallstep/step-cli). You will need to create some scripts to create and refresh the certificates, you can use `step ca renew --daemon` to have a process that automatically renews the certificate; the other option would be to use things like systemd timers.

An important thing you need to consider is that you might need to tell OpenLDAP when a certificate has been refreshed so it reloads the new certificates. Some applications have a `reload` command, or you need to send a specific signal to the server. I don't know if OpenLDAP supports anything like that. If it doesn't, the only solution would be to restart the service.

`step ca renew` has other options besides `--daemon`. For example, by combining `--signal` and `--pid` or `--pid-file`, you can send a specific signal to a process after a renewal, or you can use `--exec` to execute a specific script on renewals.

I hope this is helpful.

**Reply by ddspell** [reply]
Thank you for the detailed reply.
